library/utility/signal_handler.cc

```diff
@@ -55,7 +56,51 @@ namespace scarab
 
     signal_handler::signal_handler()
     {
-        LOGGER( slog_constr, "signal_handler constructor" );
+        //std::cerr << "signal_handler constructor" << std::endl;
+        ++signal_handler::s_ref_count;
+
+        signal_handler::handle_signals();
+    }
+
+    signal_handler::~signal_handler()
+    {
+        //std::cerr << "signal_handler destructor" << std::endl;
+        --signal_handler::s_ref_count;
+        if( signal_handler::s_ref_count == 0 )
+        {
+            signal_handler::unhandle_signals();
+        }
+    }
+
+    void signal_handler::add_cancelable( std::shared_ptr< scarab::cancelable > a_cancelable )
+    {
+        std::unique_lock< std::recursive_mutex > t_lock( s_mutex );
+        s_cancelers.insert( std::make_pair(a_cancelable.get(), cancelable_wptr_t(a_cancelable) ) );
+        return;
+    }
+
+    void signal_handler::remove_cancelable( std::shared_ptr< scarab::cancelable > a_cancelable )
+    {
+        std::unique_lock< std::recursive_mutex > t_lock( s_mutex );
+        s_cancelers.erase( a_cancelable.get() );
+        return;
+    }
+
+    void signal_handler::remove_cancelable( scarab::cancelable* a_cancelable )
+    {
+        std::unique_lock< std::recursive_mutex > t_lock( s_mutex );
+        s_cancelers.erase( a_cancelable );
+        return;
+    }
+
+    void signal_handler::handle_signals()
+    {
+        std::unique_lock< std::recursive_mutex > t_lock( s_mutex );
```

Comment:
Probably has a simple reason and I just don't understand the usage of the class well enough yet. So, why was the mutex not required before?

Reply:
Previously all of the calls to `signal()` to change how the signals are handled were done in the constructor, and un-handling signals was done in the `reset()` function (which also removed all cancelables).  This PR introduces the separate `handle()`/`unhandle()` functions, which allows the user or client code to control handling outside of the contructor and unhandling without removing cancelables.  So the mutex wasn't here before because we didn't need to use it in the constructor.